src/codegen/delete/delete_translator.cpp

```diff
@@ -0,0 +1,191 @@
+
+  if (is_owner && is_written) {
+    // if the thread is the owner of the tuple, then directly update in place.
+    LOG_TRACE("Thread is owner of the tuple");
+    transaction_manager.PerformDelete(txn, old_location);
+
+  } else {
+    bool is_ownable =
+        is_owner ||
+        transaction_manager.IsOwnable(txn, tile_group_header, tuple_id);
+
+    if (is_ownable) {
+      // if the tuple is not owned by any transaction and is visible to current
+      // transaction.
+      LOG_TRACE("Thread is not the owner of the tuple, but still visible");
+
+      bool acquire_ownership_success = is_owner ||
+                                       transaction_manager.AcquireOwnership(
+                                           txn, tile_group_header, tuple_id);
+
+      if (!acquire_ownership_success) {
+        transaction_manager.SetTransactionResult(txn, ResultType::FAILURE);
+        return false;
+      }
+      // if it is the latest version and not locked by other threads, then
+      // insert an empty version.
+      ItemPointer new_location = table->InsertEmptyVersion();
+
+      // PerformUpdate() will not be executed if the insertion failed.
+      // There is a write lock acquired, but since it is not in the write set,
+      // because we haven't yet put them into the write set.
+      // the acquired lock can't be released when the txn is aborted.
+      // the YieldOwnership() function helps us release the acquired write lock.
+      if (new_location.IsNull()) {
+        LOG_TRACE("Fail to insert new tuple. Set txn failure.");
+        if (!is_owner) {
+          // If the ownership is acquire inside this update executor, we release
+          // it here
+          transaction_manager.YieldOwnership(txn, tile_group_id, tuple_id);
+        }
+        transaction_manager.SetTransactionResult(txn, ResultType::FAILURE);
+        return false;
+      }
+      transaction_manager.PerformDelete(txn, old_location, new_location);
+
+    } else {
+      // transaction should be aborted as we cannot update the latest version.
+      LOG_TRACE("Fail to update tuple. Set txn failure.");
+      transaction_manager.SetTransactionResult(txn, ResultType::FAILURE);
+      return false;
+    }
+  }
+
+  return true;
+}
+
+const std::string &DeleteTranslator::_DeleteWrapper::GetFunctionName() {
+  static const std::string deleteWrapperFnName =
+      "_ZN7peloton7codegen16DeleteTranslator14delete_wrapperEjjPNS_"
+      "11concurrency11TransactionEPNS_7storage9DataTableE";
+  return deleteWrapperFnName;
+}
+
+llvm::Function *DeleteTranslator::_DeleteWrapper::GetFunction(
+    CodeGen &codegen) {
+  const std::string &fn_name = GetFunctionName();
+
+  // Has the function already been registered?
+  llvm::Function *llvm_fn = codegen.LookupFunction(fn_name);
+  if (llvm_fn != nullptr) {
+    return llvm_fn;
+  }
+  llvm::FunctionType *int32_type =
+      llvm::FunctionType::get(codegen.Int32Type(), false);
+  llvm::FunctionType *pointer_type =
```

Comment:
This pointer type is a char pointer. Because you need a `concurrency::Transaction *txn` and `storage::DataTable *table`, you probably need `DataTableProxy ::GetType(codegen)->GetPointerTo()`.